emission/analysis/modelling/user_label_model/util.py

```diff
@@ -0,0 +1,130 @@
+from past.utils import old_div
+import numpy
+from numpy.linalg import norm
+
+
+def load_fs(filename: str, numpy_decode: bool = True) -> dict:
+    """loads model state as a pickled object on the file system.
+    if the file is not found, returns an empty dict.
+
+    :param filename: file name to load
+    :type filename: str
+    :param numpy_decode: if part of the data is numpy encoded
+    :type numpy_decode: bool
+    :return: json object parsed, or, an empty list
+    :rtype: Dict
+    """
+    logging.debug(f"At stage: loading model")
+    try:
+        with open(filename, "r") as f:
+            contents = f.read()
+    except FileNotFoundError:
+        logging.info(f"No model found at {filename}, no prediction")
+        return {}
+    
+    try:
+        if numpy_decode:
+            # see https://jsonpickle.github.io/extensions.html
+            import jsonpickle.ext.numpy as jsonpickle_numpy
+            jsonpickle_numpy.register_handlers()
+        result = jpickle.loads(contents)
+        return result
+    except Exception as e:
+        msg = (
+            f"failure decoding stored model at {filename}, "
+            f"numpy_decode={numpy_decode}"
+        )
+        raise IOError(msg) from e
+
+
+def save_fs(filename: str, obj: object):
+    """save model state as a pickled object on the file system
+
+    :param filename: filename to write
+    :type filename: str
+    :param obj: the object to pickle + store
+    :type obj: object
+
+    """
+    try:
+        logging.debug("At stage: saving model")
+        obj_capsule = jpickle.dumps(obj)
+        with open(filename, "w") as fd:
+            fd.write(obj_capsule)
+    except Exception as e:
+        msg = f"failed writing clustering model contents to file system"
+        raise IOError(msg) from e
+
+
+def load_db(user_id: str, table: str, timestamp: Optional[int] = None) -> Dict:
+    """
+    loads a user label prediction model from a database table. 
+
+    data is assumed stored in a document database, with the structure:
+
+    { "user_id": user_id, "data": model_data }
+
+    :param user_id: user id to filter on
+    :type user_id: str
+    :param table: the table name
+    :type table: str
+    :param timestamp: optional time to 
+    :return: 
+    :rtype: Dict
+    """    
+    # build the time query if a timestamp is provided
+    time_query = lambda confirmed_trip: confirmed_trip['data']['start_ts'] >= timestamp \
+        if timestamp is not None else None
+    pass
+
+def save_db(user_id, table: str, model_data: Dict):
+    """
+    saves a user label prediction model to the database.
+
+    data is assumed stored in a document database, with the structure:
+
+    { "user_id": user_id, "data": model_data }
+
+    :param user_id: the user to store data for
+    :type user_id: object
+    :param table: the table name
+    :type table: str
+    :param model_data: the data row to store tagged by this user id
+    :type model_data: Dict
+    """
+    pass
```

Comment:
@robfitzgerald alas, storing to the DB is also a super high priority task at this point. As we work on deploying this in the NREL cloud environment (https://github.com/e-mission/e-mission-docs/issues/721), cloud services would prefer to run the modeling/inference tasks as stateless scheduled tasks. So we cannot store the model to disk, since there is no persistent disk.

We did consider storing the models to S3, but that would then require an S3 bucket approved for moderate use, which would, in turn require additional cyber approval.

Since the database is already approved for moderate use, and as a NoSQL database, mongodb supports document storage, it seemed easiest to store in the database.

Reply:
this is fine, i can work with Andrew on supporting db storage then. but i also assume we still want to support file system storage for non-NRELian users on less-restrictive deployment environments or for testing. i mean, why not, it's already in there 🤷

Reply:
ok, sounds good, i will remove the FS-related utils and we can make getting the database read/write done within this PR. as long as you're ok with letting go of backwards compatibility here. i'll reach out to you if i have questions about writing to tables (and naming conventions and schemas, but, feel free to propose those here if you have ideas beyond my initial assumptions).

Reply:
@robfitzgerald Backwards compatibility is not a huge concern right now.

At this point, the system regenerates the model from scratch every time. And when it is deployed on the NREL OpenPATH servers, it will not have any file based models to read from anyway. 

If we want to be extra nice, we could fallback to reading from file if there is no entry in the database, and then remove the fallback code after a year. That will ensure that if somebody else had been running this modeling pipeline (not aware of anybody other than CanBikeCO at this point), they can continue to use the old model to label incoming trips while the new model was being built. But there is no reason at this point to every write to the file system.

Again, not aware of anybody other than CanBikeCO who has enabled the new pipeline, so we could also just punt on the backwards compatibility if it is faster.

wrt naming conventions and schema, I have created two main collections and am storing all data in them. `Stage_timeseries` accessed via `emission.core.get_database.get_stage_timeseries()` and `Stage_analysis_timeseries` accessed via `emission.core.get_database.get_analysis_timeseries()`.

Thinking of creating similar `Stage_models_timeseries` and `Stage_results_timeseries`.

In the raw and analysis timeseries, the entries follow the same basic format with data and metadata entries, but with the fields in data being different based on `metadata.key`. This allows us to have generic metadata based queries based on `metadata.write_ts` across multiple types of data.

Not sure if the models need to be as time based, or whether they should just be UUID based. Thoughts?

Reply:
so, given your convention of maintaining immutable database entries, we could do that with the label models, and it would just mean that we take the one with the most recent write_ts entry when we need to run inference. but i'm not sure how large different models can get. i'm guessing not huge, as the datasets for each user are not huge. if that sounds sensible, we could adopt the convention of immutable records and grab the latest each time.

there's probably not a lot of value in storing all of those models, 365 models per year per user. so, the other thing we could do is assume to store these rows as mutable ones, and really only store 1 row per UUID. in that case, `metadata.write_ts` becomes less a query key and more a housekeeping field. but that would assume that UUID-based indexing is not slow in `Stage_analysis_timeseries` (or maybe motivation to create another collection configured for UUID indexing and designed/labeled for mutable DB operations).

as for the fs discussion, i'm not trying to advocate for file saving i'm just being careful that's all. if it's not helping anyone we can remove that stuff. but it's already implemented so it's not "faster" or "slower" to keep that functionality.

Reply:
I think we should start with an immutable object, but just keep the last `n` entries, with `n=3`. This should strike a good balance between robustness and disk space. We will then need a cleanup cronjob/schedule script that will go through and delete older entries.

in the long run, we should remove it because it is obsolete and I don't want to have a bunch of vestigial code sitting around either being maintained or waiting to bitrot.

in the short term, reading from DB with fallback to file seems like a bit more work than reading only from the DB.